### PR TITLE
feat: Enforce voting threshold

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #64 

- ExchangeRate will not get calculated and set unless total power of the denom's ballot is over `VoteThreshold`
- Added enblocker test for testing `VoteThreshold` allowing/disallowing denom's exchange rate to get set
- Found edge case not getting handled by `CalculateRewardFactor`

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/abci_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+	ctx sdk.Context
+	app *ojoapp.App
+}
+
+const (
+	initialPower = int64(10000000000)
+)
+
+func (s *IntegrationTestSuite) SetupTest() {
```
Comment (adamewozniak) [inline x/oracle/abci_test.go:40]: Pls add a comment on why this SetupTest is unique - also should it be public?

File: x/oracle/abci_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+func (s *IntegrationTestSuite) TestEnblockerVoteThreshold() {
+	app, ctx := s.app, s.ctx
+
+	var val1Tuples types.ExchangeRateTuples
+	var val2Tuples types.ExchangeRateTuples
+	var val1PreVotes []types.AggregateExchangeRatePrevote
+	var val2PreVotes []types.AggregateExchangeRatePrevote
+	var val1Votes []types.AggregateExchangeRateVote
+	var val2Votes []types.AggregateExchangeRateVote
```
Comment (adamewozniak) [inline x/oracle/abci_test.go:97]: Let's group these in one statement
var (...
)

**Review by adamewozniak — APPROVED** [review]
Lgtm with a couple nitpicks - let's also add a test case for when there are some denoms which have enough voting power & there are others which do not in the same endblocker

File: x/oracle/abci.go [hunk 3]
```diff
@@ -44,6 +44,10 @@ func EndBlocker(ctx sdk.Context, k keeper.Keeper) error {
 
 		// Iterate through ballots and update exchange rates; drop if not enough votes have been achieved.
 		for _, ballotDenom := range ballotDenomSlice {
+			if sdk.NewDec(ballotDenom.Ballot.Power() / 100).LTE(k.VoteThreshold(ctx)) {
```
[the remark this answers is not in the document]
Comment (rbajollari) [reply x/oracle/abci.go:47]: Added test to confirm this 👍

**Comment by adamewozniak** [comment]
@zarazan can we get a review pls?

**Review by zarazan — APPROVED** [review]
Looks good, nice test. If this file gets bigger we could pull out the IntegrationTestSuite setup to its own file.

File: x/oracle/abci_test.go (hunk 1 above)
Comment (rbajollari) [reply x/oracle/abci_test.go:42]: Its defined by the suite package so it needs the same function signature. If its changed to private it wont run before the tests.
